examples/task.hpp

```diff
@@ -23,84 +23,273 @@
+    requires std::same_as<
+        std::in_place_stop_token,
+        std::execution::stop_token_of_t<ParentPromise&>>
+  struct default_task_context_impl::awaiter_context<ParentPromise> {
+    explicit awaiter_context(
+        default_task_context_impl& self, ParentPromise& parent) {
+      self.stop_token_ = std::execution::get_stop_token(parent);
     }
   };
-  // In a base class so it can be specialized when T is void:
+
+// If the parent coroutine's stop token is unstoppable, there's no point
+// forwarding stop tokens or stop requests at all.
+template <stop_token_provider ParentPromise>
+    requires std::unstoppable_token<
+        std::execution::stop_token_of_t<ParentPromise&>>
+  struct default_task_context_impl::awaiter_context<ParentPromise> {
+    explicit awaiter_context(default_task_context_impl&, ParentPromise&) {}
+  }; 
+
+// Finally, if we don't know the parent coroutine's promise type, assume the
+// worst and save a type-erased stop callback.
+template<>
+  struct default_task_context_impl::awaiter_context<void> {
+    explicit awaiter_context(default_task_context_impl& self, auto&) {}
+
+    template <stop_token_provider ParentPromise>
+      explicit awaiter_context(
+          default_task_context_impl& self, ParentPromise& parent) {
+        // Register a callback that will request stop on this basic_task's
+        // stop_source when stop is requested on the parent coroutine's stop
+        // token.
+        using stop_token_t = std::execution::stop_token_of_t<ParentPromise&>;
+        using stop_callback_t =
+          typename stop_token_t::template callback_type<forward_stop_request>;
+        stop_callback_.emplace<stop_callback_t>(
+            std::execution::get_stop_token(parent),
+            forward_stop_request{stop_source_});
+        self.stop_token_ = stop_source_.get_token();
+      }
+
+    std::in_place_stop_source stop_source_{};
+    std::any stop_callback_{};
+  };
+
+template <class>
+  using default_task_context = default_task_context_impl;
+
+template <class Promise, class ParentPromise = void>
+  using awaiter_context_t =
+    typename Promise::template awaiter_context_t<Promise, ParentPromise>;
+
+// In a base class so it can be specialized when T is void:
+template <class T>
   struct _promise_base {
     void return_value(T value) noexcept {
       data_.template emplace<1>(std::move(value));
     }
     std::variant<std::monostate, T, std::exception_ptr> data_{};
   };
-  struct promise_type : _promise_base, std::execution::with_awaitable_senders<promise_type> {
-    task get_return_object() noexcept {
-      return task(coro::coroutine_handle<promise_type>::from_promise(*this));
+
+template<>
+  struct _promise_base<void> {
+    struct _void {};
+    void return_void() noexcept {
+      data_.template emplace<1>(_void{});
+    }
+    std::variant<std::monostate, _void, std::exception_ptr> data_{};
+  };
+
+template <class T, class Context = default_task_context<T>>
```

Comment:
I'll update the description of the PR to explain the design. But basically, the promise type inherits from `Context` to get any additional data members, and `Context::awaiter_context<ThisPromise, ParentPromise>` becomes a member of `_task_awaitable`. That `awaiter_context` member is constructed in `await_suspend` with references to both the child and parent promises, and so it can do whatever it wants. That member is destroyed first think in `await_resume`.